components/Article/Progress/index.js

```diff
@@ -0,0 +1,427 @@
+
+    this.headerHeight = () =>
+      this.mobile()
+        ? HEADER_HEIGHT_MOBILE
+        : HEADER_HEIGHT
+
+    this.initialize = (article) => {
+      const { userProgress, embeds, meta } = article
+      this.poll(userProgress)
+
+      const audioSource = meta ? meta.audioSource : undefined
+      if (embeds || audioSource) {
+        let mediaProgress = {}
+        embeds.map(embed => {
+          if (embed.userProgress) {
+            mediaProgress[embed.mediaId] = embed.userProgress.secs
+          }
+        })
+        if (audioSource &&
+          audioSource.mediaId &&
+          audioSource.userProgress &&
+          audioSource.userProgress.secs
+        ) {
+          mediaProgress[audioSource.mediaId] = audioSource.userProgress.secs
+        }
+        this.setState({ mediaProgress })
+      }
+    }
+
+    this.poll = (userProgress, pollDom) => {
+      if (!userProgress || !this.props.pollDom || (!userProgress.nodeId && !userProgress.percentage)) {
+        this.setState({ initialized: true })
+        return
+      }
+      const progressElements = this.getProgressElements()
+      const { pollRetries } = this.state
+      if (pollRetries > MAX_POLL_RETRIES) {
+        return
+      }
+      if (progressElements && progressElements.length) {
+        const { percentage, nodeId } = userProgress
+        this.restoreProgress(percentage, nodeId)
+      } else {
+        const newPollRetries = pollRetries + 1
+        this.setState({ pollRetries: newPollRetries }, () => {
+          setTimeout(() => {
+            this.poll(userProgress)
+          }, 300 * newPollRetries)
+        })
+      }
+    }
+
+    this.onScroll = () => {
+      const { isMember, article } = this.props
+      if (isMember && article) {
+        this.saveProgress(article.id)
+      }
+      this.maybeHideTopButton()
+    }
+
+    this.scrollToTop = () => {
+      window.scrollTo(0, 0)
+      this.setState({ topButtonAnimateOut: true })
+      this.resetDocumentProgress()
+    }
+
+    this.maybeHideTopButton = throttle(() => {
+      const {
+        initialized,
+        initialPageYOffset,
+        showTopButton,
+        topButtonAnimateOut
+      } = this.state
+      if (topButtonAnimateOut || !showTopButton || !initialized || !initialPageYOffset) {
+        return
+      }
+      const y = window.pageYOffset
+      const hasScrolledAway = Math.abs(y - initialPageYOffset) > SCROLLED_AWAY_PX
+      if (hasScrolledAway) {
+        this.setState({ topButtonAnimateOut: true })
+      }
+    }, 500)
+
+    this.measure = () => {
+      if (this.container) {
+        const { width, height, top } = this.container.getBoundingClientRect()
+        const cleanWidth = Math.min(width, 695)
+        if (cleanWidth !== this.state.width || height !== this.state.height || top !== this.state.top) {
+          this.setState({ width: cleanWidth, height, top })
+        }
+      }
+    }
+
+    this.measureProgress = (downwards = true) => {
```

Comment:
There is a lot happening in this method. Given that it is called on every scroll (debounced), we should look into some optimizations.